### PR TITLE
add cpuid based hypervisor vendor detection

### PR DESCRIPTION
**Pull request by SvenDowideit**
code lifted from https://github.com/zcalusic/sysinfo and https://people.redhat.com/~rjones/virt-what/

I'd like DMI support, but that's a future PR (today I need vmware detection in RancherOS :) )

### CODE REVIEW

**Comment by SvenDowideit** [comment]
I'm not 100% sure adding DMI to this library is the right way to go - we'll see :)

**Comment by SvenDowideit** [comment]
@klauspost looks like I don't understand your test harness - can you help with the failure?

File: cpuid.go [hunk 1]
```diff
@@ -554,6 +559,54 @@ func maxFunctionID() uint32 {
...
 
+func isHypervisorActive() bool {
+	_, _, i, _ := cpuid(0x1)
+	return i&(1<<31) != 0
+}
+
+func getHypervisorCpuid(ax uint32) string {
+	a, b, c, d := cpuid(ax)
+	var info [4]uint32 = [4]uint32{a, b, c, d}
+	name := strings.TrimRight(string((*[12]byte)(unsafe.Pointer(&info[1]))[:]), "\000")
```
Comment (klauspost) [inline cpuid.go]: Instead of unsafe, use `valAsString` function to convert it.

File: cpuid.go [hunk 2]
```diff
@@ -554,6 +559,54 @@ func maxFunctionID() uint32 {
...
+
+	if hv := getHypervisorCpuid(0x40000000); hv != "" {
+		return hv
+	}
+	return ""
+}
+
+// https://en.wikipedia.org/wiki/CPUID#EAX.3D0:_Get_vendor_ID
+var hvmap = map[string]string{
+	"bhyve bhyve ": "bhyve",
```
Comment (klauspost) [inline cpuid.go]: What does these values relate to? Is there a standard, or are they just arbitrarily chosen?

File: cpuid.go [hunk 3]
```diff
@@ -554,6 +559,54 @@ func maxFunctionID() uint32 {
...
+		return ""
+	}
+	// KVM has been caught to move its real signature to this leaf, and put something completely different in the
+	// standard location. So this leaf must be checked first.
+	// Sven removed it - in one test system, this leaf returns garbage :(
+	//if hv := getHypervisorCpuid(0x40000100); hv != "" {
+	//	return hv
+	//}
+
+	if hv := getHypervisorCpuid(0x40000000); hv != "" {
```
Comment (klauspost) [inline cpuid.go]: Just `return getHypervisorCpuid(0x40000000)` - the remaining code does nothing.

Not sure if you should check availability of it somehow.

File: cpuid.go [hunk 4]
```diff
@@ -554,6 +559,54 @@ func maxFunctionID() uint32 {
 	return a
 }
 
+func isHypervisorActive() bool {
+	_, _, i, _ := cpuid(0x1)
```
Comment (klauspost) [inline cpuid.go]:
```
	if maxFunctionID() < 0x1 {
		return false
	}
```
This is why tests fail.

File: cpuid.go (hunk 2 above)
Comment (SvenDowideit) [inline cpuid.go]: I get the feeling that hypervisor projects just choose something, and if they get popular, these become "well known"

File: cpuid.go (hunk 4 above)
Comment (klauspost) [inline cpuid.go]: The code above should be inserted, to check if the function is available on the CPU. I believe if you insert this, it should work.

File: cpuid.go (hunk 2 above)
Comment (klauspost) [inline cpuid.go]: OK, lets add these as exported constants. At least then people have something to compare it to, like:
```Go
const (
  HypervisorBhyve = "bhyve"
  HypervisorKvm = "kvm"
  HypervisorHyperV = "hyperv"
  HypervisorVmware = "vmware"
  HypervisorXen = "xenhvm"
)
```

File: mockcpu_test.go [hunk 5]
```diff
@@ -103,6 +103,10 @@ func mockCPU(def []byte) func() {
 				return 0, 0, 0, 0
 			}
 		}
+		if op == 0x1 {
```
Comment (klauspost) [inline mockcpu_test.go]: This shouldn't really be needed. If it is, I suspect that there may be some compatibility issues.

**Comment by ribasushi** [comment]
@klauspost was this PR fully merged or are there bits that were left out? Just wondering...

**Comment by klauspost** [comment]
Closing for now. Feel free to reopen.

**Pull request closed without merging**